### PR TITLE
Finetuning with weights in bfloat16

### PR DESCRIPTION
**Pull request by awaelchli**
The memory consumption is now ~20GB compared to before at ~38GB. 
The training iteration speed-up ~1.5x-2x.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -136,7 +136,7 @@ We provide a simple training scripts in `finetune_lora.py` and `finetune_adapter
    ```
 
 It is expected that you have downloaded the pretrained weights as described above.
-The finetuning requires a GPU with 40 GB memory (A100).
+The finetuning requires a GPU with ~24 GB memory (GTX 3090).
 Coming soon: LoRA + quantization for training on a consumer-grade GPU!
```
Comment (lantiga) [inline README.md:140]:
```suggestion
```

File: README.md [hunk 2]
```diff
@@ -136,7 +136,7 @@ We provide a simple training scripts in `finetune_lora.py` and `finetune_adapter
    ```
 
 It is expected that you have downloaded the pretrained weights as described above.
-The finetuning requires a GPU with 40 GB memory (A100).
+The finetuning requires a GPU with ~24 GB memory (GTX 3090).
```
Comment (lantiga) [inline README.md:139]:
```suggestion
Finetuning using LLaMA-Adapter requires a GPU with ~24 GB memory (GTX 3090).
```

**Review by lantiga — APPROVED** [review]
A few suggested changes

**Comment by lantiga** [comment]
Does memory fit a 3090 even in the LoRA case?

**Comment by awaelchli** [comment]
It's tight but yeah it could work:

image

I can double-check on our 3090 machine.

**Comment by lantiga** [comment]
better check IMO

**Comment by awaelchli** [comment]
Attempted to test it but got hold up by this problem: #101

**Comment by awaelchli** [comment]
The finetuning fits into the 3090:

image

I had to install pytorch nightly to get around the issue #101 though. Perhaps we should hold off merging this? Or we could say this requires pytorch nightly. Or we could investigate whether changing the implementation can avoid the error.

**Comment by lantiga** [comment]
Great, I would merge and say that 3090 requires nightly

then we’ll investigate complex again

**Comment by lantiga** [comment]
Amazing, let’s merge!

**Pull request merged**